File: PYTHON/vehiclemodels/sim/actuators/aero.py
"""Aerodynamic drag and downforce helpers."""
from __future__ import annotations

from dataclasses import dataclass
import math


@dataclass
class AeroConfig:
    drag_coefficient: float
    downforce_coefficient: float


class AeroModel:
    def __init__(self, config: AeroConfig) -> None:
        self.config = config

    def drag_force(self, speed: float) -> float:
        coeff = self.config.drag_coefficient
        return -math.copysign(coeff * speed * speed, speed) if speed else 0.0

    def downforce(self, speed: float) -> float:
        coeff = self.config.downforce_coefficient
        if coeff == 0:
            return 0.0
        return abs(coeff) * speed * speed

File: lib/controllers/longitudinal/aero.cpp
#include "controllers/longitudinal/aero.hpp"

#include <cmath>
#include <stdexcept>

namespace velox::controllers::longitudinal {

void AeroConfig::validate() const
{
    if (!std::isfinite(drag_coefficient)) {
        throw std::invalid_argument("aero.drag_coefficient must be finite");
    }
    if (!std::isfinite(downforce_coefficient)) {
        throw std::invalid_argument("aero.downforce_coefficient must be finite");
    }
    if (drag_coefficient < 0.0) {
        throw std::invalid_argument("aero.drag_coefficient cannot be negative");
    }
}

AeroModel::AeroModel(AeroConfig config)
    : config_(config)
{
    config_.validate();
}

double AeroModel::drag_force(double speed) const noexcept
{
    if (speed == 0.0) {
        return 0.0;
    }
    const double coeff     = config_.drag_coefficient;
    const double magnitude = coeff * speed * speed;
    return -std::copysign(magnitude, speed);
}

double AeroModel::downforce(double speed) const noexcept
{
    if (config_.downforce_coefficient == 0.0 || speed == 0.0) {
        return 0.0;
    }
    const double coeff = std::abs(config_.downforce_coefficient);
    return coeff * speed * speed;
}

} // namespace velox::controllers::longitudinal
